core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowRewriter.scala

```diff
@@ -0,0 +1,424 @@
+        r1NewBreakpoints += breakpoint
+      }
+    })
+  }
+
+  private def r2(opID: String): Unit = {
+    val op = workflowDAG.getOperator(opID)
+    if (isCacheEnabled(op) && isCacheValid(op)) {
+      val cacheSourceOp = getCacheSourceOperator(op)
+      r2NewOps += cacheSourceOp
+      workflowDAG.jgraphtDag
+        .outgoingEdgesOf(opID)
+        .forEach(link => {
+          val src = OperatorPort(cacheSourceOp.operatorID, link.origin.portOrdinal)
+          val dest = link.destination
+          r2NewLinks += OperatorLink(src, dest)
+        })
+      workflowInfo.breakpoints.foreach(breakpoint => {
+        if (breakpoint.operatorID.equals(opID)) {
+          r2NewBreakpoints += BreakpointInfo(cacheSourceOp.operatorID, breakpoint.breakpoint)
+        }
+      })
+    } else {
+      r2NewOps += op
+      workflowDAG.jgraphtDag.outgoingEdgesOf(opID).forEach(link => r2NewLinks.+=(link))
+      workflowInfo.breakpoints.foreach(breakpoint => {
+        if (breakpoint.operatorID.equals(op.operatorID)) {
+          r2NewBreakpoints += breakpoint
+        }
+      })
+      workflowDAG
+        .getUpstream(opID)
+        .map(_.operatorID)
+        .foreach(id => {
+          if (!r2OpIDQue.contains(id)) {
+            r2OpIDQue += id
+          }
+        })
+    }
+  }
+
+  def cacheStatusUpdate(): mutable.Set[String] = {
+
+    val invalidSet = mutable.HashSet[String]()
+
+    def invalidateOperatorCache(id: String): Unit = {
+      if (cachedOperatorDescriptors.contains(id)) {
+        cachedOperatorDescriptors.remove(id)
+        cacheSinkOperatorDescriptors.remove(id)
+        cacheSourceOperatorDescriptors.remove(id)
+        invalidSet += id
+      }
+      logger.info("Operator {} cache invalidated.", id)
+    }
+
+    def invalidateOperatorCacheRecursively(opID: String): Unit = {
+      invalidateOperatorCache(opID)
+      workflowDAG
+        .getDownstream(opID)
+        .foreach(desc => {
+          invalidateOperatorCacheRecursively(desc.operatorID)
+        })
+    }
+
+    def isUpdated(opID: String): Boolean = {
+      if (!operatorRecord.contains(opID)) {
+        operatorRecord += ((opID, getWorkflowVertex(workflowDAG.getOperator(opID))))
+        logger.info("Vertex {} is not recorded.", operatorRecord(opID))
+        true
+      } else if (workflowInfo.cachedOperatorIDs.contains(opID)) {
+        !operatorRecord(opID).equals(getWorkflowVertex(workflowDAG.getOperator(opID)))
+      } else {
+        val vertex = getWorkflowVertex(workflowDAG.getOperator(opID))
+        if (!operatorRecord(opID).equals(vertex)) {
+          operatorRecord(opID) = vertex
+          logger.info("Vertex {} is updated.", operatorRecord(opID))
+          true
+        } else if (cachedOperatorDescriptors.contains(opID)) {
+          !workflowInfo.cachedOperatorIDs.contains(opID)
+        } else {
+          logger.info("Operator: {} is not updated.", operatorRecord(opID))
+          false
+        }
+      }
```

Comment:
I believe this is a core function to check whether the result is updated (again the function name is not that clearly readable and meaningful). From reading the code, I know there are 5 branches, but I am not very sure what does each branch mean. Please add some comments.